### ISSUE TITLE
Automatically handling Docker networks

### ISSUE THREAD

**Issue by michaellopez** [codekitchen/dinghy#218]
Does Dinghy actively automatically join Docker networks?

I mean, when I start up my projects with `docker-compose up -d` with a version 2 compose-file, without any specific networks defined, Docker automatically creates a bridged network for that particular project. As far as I understand, the Dinghy proxy needs to join that network to be able to communicate with the containers in that network to be able to do its job properly. This is exactly what is happening on my machine and it is working fine, I'm just wondering if this is done by Dinghy intentionally?

What I'm getting at is that on `docker-compose down` I always get
```sh
Removing network somename_default
ERROR: network somename_default has active endpoints
```
That so called `active endpoint` is the Dinghy proxy that joined the network (automagically?) when the project was started. So to really clean up properly (basically running `docker-compose down -v`) I first have to disconnect the Dinghy proxy from the networks I want to clean up. Once disconnected, clean up works fine.

So my feature request is really: can Dinghy listen for a Docker event of `REMOVE` or similar (just the container with `VIRTUAL_HOST` is good enough) and when it sees that, disconnect itself from any non-default network? It would speed up our workflow a bit.

Let me know if this is a desired functionality and point me in the right direction of the code and maybe I can send a PR for this.

**Comment by codekitchen** [comment]
Yep, that's all accurate. The code that joins the networks is at https://github.com/codekitchen/dinghy-http-proxy/blob/master/join-networks.go

I'm a little surprised this hasn't come up before, actually.

Note that code doesn't run continuously listening to events, it gets run by `docker-gen` when containers are created/removed. We might be able to get away with just checking for networks we've joined that no longer have any exposed containers, and removing ourself from those. That'll depend on if our code can run between the time that compose removes the last container and removes the network, though, so it might be touchy.

**Comment by ryansch** [comment]
Here's some ruby code I wrote to deal with the docker event stream:
https://github.com/outstand/consul_bridge/blob/master/lib/consul_bridge/monitor_docker_events.rb

**Comment by michaellopez** [comment]
@codekitchen It might be that not that many use Compose V2 syntax. V1 syntax does not have this problem, as it does not create networks as V2 does...I'll see if I can do something about this after the holidays...

**Comment by ryansch** [comment]
I use v2 but I almost never remember to run `docker-compose down`...

**Comment by codekitchen** [comment]
Yeah we added the network joiner code specifically to make the proxy work with v2 syntax, since that always creates a network. I'm in the same boat, most all the apps at my company are migrated over to v2 syntax at this point, but none of us have really made `docker-compose down` part of our workflow.

**Comment by codekitchen** [comment]
I had some time to take a crack at this, and so far it seems to work just fine. There's enough lag between docker stopping the containers and dropping the network, that the proxy has time to leave the network as well. Of course, there's sadly no way to guarantee the timing.

I've released this as part of dinghy-http-proxy:2.5.5, if you are on the latest version of dinghy you can force an upgrade by running `docker pull codekitchen/dinghy-http-proxy:2.5` and then `dinghy restart`.

If the timing does end up being an issue, we might have to just suggest that people run `docker-compose stop` before `docker-compose down`, since the `stop` will trigger the proxy to leave the network.

**Comment by michaellopez** [comment]
@codekitchen wow, I appreciate you taking the time to have a look at this. I will admit, looking at the patch, I realised my Go fu is not on the level needed to solve this :) I will play with this asap and let you know if I find any hiccups. Again, thanks, awesome maintainer. 🍰

**Comment by michaellopez** [comment]
@codekitchen I'm not having any luck forcing an upgrade of the proxy inside Dinghy. I ran the commands you posted, but I still get `2.5.3` even after removing the `2.5.3` image. It just redownloads it. I tracked it down to https://github.com/codekitchen/dinghy/commit/49f0d8481a7d7aa21df56986593cdb7b7f4a28d9 being required for your instructions to work. That's not released yet. I'm on Dinghy 4.4.3. So for now, we should all just sit tight until the next Dinghy version release (or be adventurous and install HEAD or something, but since this isn't a blocker I'll wait) Just wanted to share in case any future googlers come around looking for help.
